interval/window.go

```diff
@@ -0,0 +1,310 @@
+// GetLatestBounds returns the bounds for the latest window bounds
+// that contains the given time t.  For underlapping windows that
+// do not contain time t, the window directly before time t will be returned.
+func (w Window) GetLatestBounds(t values.Time) Bounds {
+	// Get the latest index that should contain the time t
+	index := w.lastIndex(t)
+	// Construct the bounds from the index
+	start := w.zero.Add(w.every.Mul(index))
+	b := Bounds{
+		start: start,
+		stop:  start.Add(w.period),
+		index: index,
+	}
+	// If the period is negative its possible future bounds can still contain this point
+	if w.period.IsNegative() {
+		// swap start and stop since the period was negative
+		b.start, b.stop = b.stop, b.start
+		// If period is NOT mixed we can do a direct calculation
+		// to determine how far into the future a bounds may be found.
+		if !w.period.IsMixed() {
+			// Since its not mixed we can adjust the index closer based
+			// on how many windows a period can span
+			var period, every int64
+			if w.every.MonthsOnly() {
+				every = w.every.Months()
+				period = w.period.Months()
+			} else {
+				every = w.every.Nanoseconds()
+				period = w.period.Nanoseconds()
+			}
+			if period > every {
+				indexDelta := period / every
+				index += int(indexDelta)
+			}
+		}
+		// Now do a direct search
+		next := w.NextBounds(b)
+		for next.Contains(t) {
+			b = next
+			next = w.NextBounds(next)
+		}
+	}
+	return b
+}
+
+// GetOverlappingBounds returns a slice of bounds that overlaps the input bounds.
+// The returned set of bounds are ordered by decreasing time.
+func (w Window) GetOverlappingBounds(start, stop values.Time) []Bounds {
+	bounds := Bounds{
+		start: start,
+		stop:  stop,
+	}
+	if bounds.IsEmpty() {
+		return []Bounds{}
+	}
+
+	// Estimate the number of windows by using a rough approximation.
+	count := (bounds.Length().Duration() / w.every.Duration()) + (w.period.Duration() / w.every.Duration())
+	bs := make([]Bounds, 0, count)
+
+	curr := w.GetLatestBounds(stop)
+	for curr.stop > start {
+		if curr.Overlaps(bounds) {
+			bs = append(bs, curr)
+		}
+		curr = w.PrevBounds(curr)
+	}
+
+	return bs
+}
+
+// NextBounds returns the next boundary in sequence from the given boundary.
+func (w Window) NextBounds(b Bounds) Bounds {
+	index := b.index + 1
+	start := w.zero.Add(w.every.Mul(index))
+	stop := start.Add(w.period)
+	if w.period.IsNegative() {
+		start, stop = stop, start
+	}
+	return Bounds{
+		start: start,
+		stop:  stop,
+		index: index,
+	}
+}
+
+// PrevBounds returns the previous boundary in sequence from the given boundary.
+func (w Window) PrevBounds(b Bounds) Bounds {
+	index := b.index - 1
+	start := w.zero.Add(w.every.Mul(index))
+	stop := start.Add(w.period)
+	if w.period.IsNegative() {
+		start, stop = stop, start
+	}
+	return Bounds{
+		start: start,
+		stop:  stop,
+		index: index,
+	}
+}
+
+// lastIndex will compute the index of the last bounds to contain t
+func (w Window) lastIndex(t values.Time) int {
```

Comment:
It would be helpful if this method contained some kind of caching mechanism for reducing the search space based on repeated queries into the window. Generally, the window will be asked for the first or last boundary for points that are all fairly close to each other. Unfortunately, the most complex calculation is usually the first one as you are trying to index into a value that is, by its nature, potentially `O(n)`. We can speed that up through math and indexing certain values, but I think it will still fundamentally be slow in certain situations.

From my experience, this function will be called at least once for every point. Each time I've written these functions, I've thought about using the previous bounds and just retrieving the next boundary, but there's no guarantee that it won't degenerate into an `O(n)` scan for sparse data.

I'm just spitballing at the moment with this.